Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -362,6 +499,39 @@ bool TextureMapperLayer::shouldBlend() const
     return m_currentOpacity < 1;
 }
 
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    bool isLeafOf3DScene = !m_state.preserves3D && (m_parent && m_parent->preserves3D());
+    bool hasPerspective = m_layerTransforms.localTransform.hasPerspective() && m_layerTransforms.localTransform.m34() != -1.f;
+    if ((isLeafOf3DScene || hasPerspective) && !m_children.isEmpty() && has3DLocalTransform())
```

Comment:
No, there's not need to flatten to intermediate buffer if there's no 3D transform in leaf node. In case no 3D leaf can be rendered to parent surface. Also perspective with 3D transform need to propagate perspective and 3D transform to all of its child layers so it behaves in a similar manner as 3D leaf. There's test scenario for perspective case in https://bugs.webkit.org/show_bug.cgi?id=281113